ext/dom/html_document.c

```diff
@@ -0,0 +1,1047 @@
+	dom_decoding_encoding_ctx_init(&decoding_encoding_ctx);
+	if (override_encoding != NULL) {
+		const lxb_encoding_data_t *encoding_data = lxb_encoding_data_by_name((const lxb_char_t *) override_encoding, override_encoding_len);
+		if (!encoding_data) {
+			zend_argument_value_error(3, "must be a valid document encoding");
+			RETURN_THROWS();
+		}
+		dom_setup_parser_encoding_manually(buf_ref, encoding_data, &decoding_encoding_ctx, &application_data);
+	} else {
+		dom_setup_parser_encoding_implicitly(&buf_ref, &source_len, &decoding_encoding_ctx, &application_data);
+	}
+
+	lxb_html_document_t *document = lxb_html_document_create();
+	if (UNEXPECTED(document == NULL)) {
+		goto fail_oom;
+	}
+
+	lxb_status_t lexbor_status = lxb_html_document_parse_chunk_begin(document);
+	if (UNEXPECTED(lexbor_status != LXB_STATUS_OK)) {
+		goto fail_oom;
+	}
+
+	lxb_html_parser_t *parser = document->dom_document.parser;
+
+	while (source_len > 0) {
+		size_t chunk_size = source_len;
+		if (chunk_size > sizeof(decoding_encoding_ctx.encoding_output) / sizeof(lxb_char_t)) {
+			chunk_size = sizeof(decoding_encoding_ctx.encoding_output) / sizeof(lxb_char_t);
+		}
+		source_len -= chunk_size;
+
+		const lxb_char_t *buf_end = buf_ref + chunk_size;
+		bool result = dom_parse_decode_encode_step(&ctx, document, parser, &buf_ref, buf_end, &decoding_encoding_ctx, &tokenizer_error_offset, &tree_error_offset);
+		if (!result) {
+			goto fail_oom;
+		}
+	}
+
+	if (!dom_parse_decode_encode_finish(&ctx, document, parser, &decoding_encoding_ctx, &tokenizer_error_offset, &tree_error_offset)) {
+		goto fail_oom;
+	}
+
+	lexbor_status = lxb_html_document_parse_chunk_end(document);
+	if (lexbor_status != LXB_STATUS_OK) {
+		goto fail_oom;
+	}
+
+	xmlDocPtr lxml_doc;
+	lexbor_libxml2_bridge_status bridge_status = lexbor_libxml2_bridge_convert_document(document, &lxml_doc, options & XML_PARSE_COMPACT, !(options & DOM_HTML_NO_DEFAULT_NS));
+	lexbor_libxml2_bridge_copy_observations(parser->tree, &ctx.observations);
+	if (UNEXPECTED(bridge_status != LEXBOR_LIBXML2_BRIDGE_STATUS_OK)) {
+		php_libxml_ctx_error(NULL, "%s in %s", dom_lexbor_libxml2_bridge_status_code_to_string(bridge_status), application_data.input_name);
+		lxb_html_document_destroy(document);
+		RETURN_FALSE;
+	}
+	lxb_html_document_destroy(document);
+
+	dom_post_process_html5_loading(lxml_doc, options, &ctx.observations);
+
+	if (decoding_encoding_ctx.decode_data) {
+		lxml_doc->encoding = xmlStrdup((const xmlChar *) decoding_encoding_ctx.decode_data->name);
+	} else {
+		lxml_doc->encoding = xmlStrdup((const xmlChar *) "UTF-8");
+	}
+
+	dom_object *intern = php_dom_instantiate_object_helper(return_value, dom_html_document_class_entry, (xmlNodePtr) lxml_doc, NULL);
+	intern->document->is_modern_api_class = true;
+	return;
+
+fail_oom:
+	lxb_html_document_destroy(document);
+	php_dom_throw_error(INVALID_STATE_ERR, 1);
+	RETURN_THROWS();
+}
+
+PHP_METHOD(DOM_HTMLDocument, createFromFile)
```

Comment:
For XML the `createFromFile()` and `createFromString()` are merged, is this not possible for HTML?

Reply:
It is possible, but I think that will make the code more complicated

Reply:
Okay, just wanted to make sure it wasn't just an oversight. I didn't check what the differences are.